### PR TITLE
MAINT Migrates rrelu_with_noise from THC to ATen on Cuda

### LINKED ISSUE #24618: Migrate `rrelu_with_noise` and `rrelu_with_noise_` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### PR DESCRIPTION
**Pull request by thomasjpfan**
Fixes #24618
Related to #24507

<details><summary>Benchmark script:</summary>

```py
import torch
import torch.nn as nn
import time

torch.manual_seed(0)
def _time():
    torch.cuda.synchronize()
    return time.time()

device = "cuda"
m = nn.RReLU().cuda()

for n in [100, 10_000, 100_000]:
    fwd_t = 0
    bwd_t = 0
    input = torch.randn(128, n, device=device)
    grad_output = torch.ones(128, n, device=device)
    for i in range(10000):
        t1 = _time()
        output = m(input)
        t2 = _time()
        fwd_t = fwd_t + (t2 -t1)
    fwd_avg = fwd_t / 10000 * 1000
    print(f"input size(128, {n}) forward time is {fwd_avg:.2f} (ms)")
```

</details>

#### This PR

```
input size(128, 100) forward time is 0.01 (ms)
input size(128, 10000) forward time is 0.06 (ms)
input size(128, 100000) forward time is 0.54 (ms)
```

#### On master

```
input size(128, 100) forward time is 0.01 (ms)
input size(128, 10000) forward time is 0.08 (ms)
input size(128, 100000) forward time is 0.66 (ms)
```

### CODE REVIEW

**Comment by ezyang** [comment]
removed myself

File: aten/src/ATen/native/cuda/Activation.cu [hunk 1]
```diff
@@ -243,6 +247,160 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
...
 }
 
+// -----------------------------------
+// rrelu
+// -----------------------------------
+template<typename scalar_t>
+inline scalar_t __device__ curand_uniform_type(curandStatePhilox4_32_10_t *state);
+
+template <>
+inline THHalf __device__ curand_uniform_type<THHalf>(curandStatePhilox4_32_10_t *state) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Activation.cu:257]: Don't use legacy THHalf type, use at::Half instead. There are implicit conversions between at::Half and float, so ScalarConvert is not necessary

File: aten/src/ATen/native/cuda/Activation.cu [hunk 2]
```diff
@@ -243,6 +247,160 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
...
+  auto rand = curand_uniform4(state);
+  return ScalarConvert<float, THHalf>::to(rand.x);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Activation.cu:259]: using only .x out of 4 generated numbers is wasteful, you can have an unroll loop in the kernel that would use all the values, you can take a look e.g. at the non-vectorized `fused_dropout_kernel` in Dropout.cu

File: aten/src/ATen/native/cuda/Activation.cu [hunk 3]
```diff
@@ -243,6 +247,160 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
...
+              idx,
+              std::get<1>(seeds),
+              &state);
+
+  CUDA_KERNEL_LOOP(i, n)
+  {
+    if (input[i] <= 0)
+    {
+      scalar_t r = curand_uniform_type<scalar_t>(&state);
+      r = ScalarConvert<double, scalar_t>::to(r * (b - a) + a);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Activation.cu:296]: having `double` is usually perf penalty, it should be `scalar_t` or at most `accscalar_t`

File: aten/src/ATen/native/cuda/Activation.cu [hunk 4]
```diff
@@ -243,6 +247,160 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
...
+      output[i] = input[i] * r;
+      noise[i] = r;
+    }
+    else
+    {
+      output[i] = input[i];
+      noise[i] = ScalarConvert<int, scalar_t>::to(1);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Activation.cu:303]: No ScalarConvert please

File: aten/src/ATen/native/cuda/Activation.cu [hunk 5]
```diff
@@ -243,6 +247,160 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
...
+  int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  curandStatePhilox4_32_10_t state;
+  curand_init(std::get<0>(seeds),
+              idx,
+              std::get<1>(seeds),
+              &state);
+
+  CUDA_KERNEL_LOOP(i, n)
+  {
+    if (input[i] <= 0)
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Activation.cu:293]: to avoid warp divergence, you should generate randoms for every input, and then only diverge on fast operations like computing output and noise

**Comment by thomasjpfan** [comment]
Thank you for the review @ngimel ! I updated the PR to use unrolling. I ran the following benchmark:


<details><summary>Benchmark script:</summary>

```python
import torch
import torch.nn as nn
import time

torch.manual_seed(0)
def _time():
    torch.cuda.synchronize()
    return time.time()

device = "cuda"
m = nn.RReLU().cuda()
n_runs = 1_000

for n in [10_000, 100_000, 1_000_000]:
    fwd_t = 0
    bwd_t = 0
    input = torch.randn(128, n, device=device)
    grad_output = torch.ones(128, n, device=device)
    for i in range(n_runs):
        t1 = _time()
        output = m(input)
        t2 = _time()
        fwd_t = fwd_t + (t2 -t1)
    fwd_avg = fwd_t / n_runs * 1000
    print(f"input size(128, {n}) forward time is {fwd_avg:.2f} (ms)")

```

</details>

### Results from benchmark:

#### This PR

```
input size(128, 10000) forward time is 0.06 (ms)
input size(128, 100000) forward time is 0.43 (ms)
input size(128, 1000000) forward time is 4.17 (ms)
```

#### On master

```
input size(128, 10000) forward time is 0.09 (ms)
input size(128, 100000) forward time is 0.69 (ms)
input size(128, 1000000) forward time is 6.66 (ms)
```

File: aten/src/ATen/native/cuda/Activation.cu [hunk 6]
```diff
@@ -243,6 +247,170 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
...
+              idx,
+              std::get<1>(seeds),
+              &state);
+
+  int grid_stride = blockDim.x * gridDim.x * unroll_factor;
+  int rounded_size = ((numel - 1) / grid_stride + 1) * grid_stride;
+  double range = upper - lower;
+
+  for (int linear_index = idx; linear_index < rounded_size; linear_index += grid_stride) {
+    auto rand = random_func(&state);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Activation.cu:279]: Can you please add static assert here that sizeof(rand)/sizeof(rand.x) == unroll_factor? Otherwise your (&rand.x)[ii] access is unsafe.

File: aten/src/ATen/native/cuda/Activation.cu [hunk 7]
```diff
@@ -243,6 +247,170 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
...
+    const Scalar& upper,
+    bool training,
+    c10::optional<Generator> generator,
+    Tensor& output) {
+  TensorArg self_arg{self, "self", 1}, noise_arg{noise, "noise", 2},
+      output_arg{output, "output", 3};
+  checkAllSameGPU("rrelu_with_noise_out_cuda", {self_arg, noise_arg, output_arg});
+
+  auto input = self.contiguous();
+  auto noise_ = noise.contiguous();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Activation.cu:374]: rrelu_with_noise_out_cuda is a user facing function, which means that `output` can also be discontiguous here.

File: aten/src/ATen/native/cuda/Activation.cu [hunk 8]
```diff
@@ -243,6 +247,170 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
...
+
+  if (training) {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+        self.scalar_type(), "rrelu_with_noise_out_cuda", [&] {
+          _rrelu_with_noise_cuda_train<scalar_t>(
+              output, input, noise_, lower, upper, generator);
+        });
+  } else {
+    auto lower_tensor = scalar_to_tensor(lower);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Activation.cu:383]: you don't need to convert Scalar to tensor here, instead convert to regular type (using `.to<double>`) and negative_slope back to Scalar

**Review by ngimel — COMMENTED** [review]
Thanks, this looks good, I left minor comments.

File: aten/src/ATen/native/cuda/Activation.cu [hunk 9]
```diff
@@ -277,6 +277,10 @@ __global__ void rrelu_with_noise_cuda_kernel(
 
   for (int linear_index = idx; linear_index < rounded_size; linear_index += grid_stride) {
     auto rand = random_func(&state);
+
+    // ensure that (&rand.x)[ii] is safe
+    CUDA_KERNEL_ASSERT(sizeof(rand)/sizeof(rand.x) == unroll_factor);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Activation.cu:282]: it should be static_assert (to be done at compile time), not runtime assert.

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
Can you please try rebasing, to get CI signal?
